### PR TITLE
Added callable check for www credentials.

### PR DESCRIPTION
**Pull request by macdems**
This adds a custom callable check for www credentials. The main use-case is user verification against some external database.

### CODE REVIEW

File: master/docs/manual/cfg-www.rst [hunk 1]
```diff
@@ -266,6 +266,28 @@ The available classes are described here:
...
+            if user == 'snow' and password == 'white':
+                return True
+            else:
+                return False
+
+        from buildbot.plugins import util
+        c['www'] = {
+            # ...
+            'auth': util.CallableAuth(check_credentials),
+        }
```
Comment (tardyp) [inline master/docs/manual/cfg-www.rst]: I usually try to ban callables as much as I can in new Buildbot APIs.

I would prefer an subclassing helper.

```
     class MyAuth(util.CustomAuth):
           def check_credentials(user, password):
                     [...]

     c['www']['auth'] = MyAuth()
```

Implementation could look like this:

```
@implementer(ICredentialsChecker)
class CustomAuth(TwistedICredAuthBase):
    credentialInterfaces = IUsernamePassword
    def requestAvatarId(self, cred):
        if self.callable(cred.username, cred.password):
            return defer.succeed(cred.username)
        return defer.fail(UnauthorizedLogin())

    def __init__(self, **kwargs):
        TwistedICredAuthBase.__init__(
            self,
            [BasicCredentialFactory(b"buildbot")],
            [self],
            **kwargs)
```

File: master/docs/manual/cfg-www.rst [hunk 2]
```diff
@@ -266,6 +266,24 @@ The available classes are described here:
             'auth': util.UserPasswordAuth({"homer": "doh!"}),
         }
 
+.. py:class:: buildbot.www.auth.CustomAuth()
+
+    This authentication class means to be overridden with a custom ``check_credentials`` method that gets username and password
+    as arguments and check if the user can login. You may use it e.g. to check the credentials against an external database or file.
+
+    ::
+        from buildbot.plugins import util
```
Comment (tardyp) [inline master/docs/manual/cfg-www.rst]: for this to work, you have to add CustomAuth into setup.py

**Comment by macdems** [comment]
Can you tell me why the automatic tests failed? The error messages seem to have nothing to do with my changes...

**Comment by tardyp** [comment]
That is an instability. Sorry for overlook

**Comment by macdems** [comment]
Thanks :)

**Pull request merged**